OAK-11603: lucene 4.x fuzzy queries don't work in Elastic

fabriziofortino opened the pull request.

fabriziofortino committed:
OAK-11603: lucene 4.x fuzzy queries don't work in Elastic

alagodasii commented on oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java:
```diff
@@ -908,6 +914,66 @@ private static QueryStringQuery.Builder fullTextQuery(String text, String fieldN
+     * @param text The input query string containing fuzzy or proximity queries.
+     * @return A query string where old fuzzy syntax is converted to the new format.
+     */
+    private String convertFuzzyQuery(String text) {
```
This method perfectly meets criteria for `utility`, because it has no dependencies - just a pure logic. How about extracting it to some utility class? Doing so, you would be able to test it more robustly without boilerplate setup of higher level tests.

fabriziofortino replied:
In general, I prefer to move these kind of methods in a utility class when are used in multiple classes. It's not the case here since it's used by a single method.

nfsantos replied:
But maybe this method and the method that calls it can be made static?

alagodasii commented:
Tests in this repo fail very frequently. Here again. Should it be a warning for us or we can ignore it?

nfsantos commented on oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java:
```diff
@@ -908,6 +914,66 @@ private static QueryStringQuery.Builder fullTextQuery(String text, String fieldN
+        Matcher oldMatcher = OLD_FUZZY_PATTERN.matcher(text);
+        StringBuilder result = new StringBuilder();
+
+        while (oldMatcher.find()) {
```
If the query does not contain a fuzzy expression (maybe the `~` in the query was quoted and is not part of a fuzzy expression or someone in the future removed the check for `~` done before calling this method), this method will always create a copy of the argument. We could instead return the argument unchanged and avoid creating a new copy if it does not match a fuzzy query.

nfsantos commented on oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java:
```diff
@@ -908,6 +914,66 @@ private static QueryStringQuery.Builder fullTextQuery(String text, String fieldN
+                    editDistance = 1;
+                }
+            } catch (NumberFormatException e) {
+                LOG.warn("Invalid fuzzy value: {}, using default edit distance of 2", fuzzyValue);
```
Should we log the query as well? Otherwise we do not have the context to understand where the this error occurred.

nfsantos reviewed (changes requested):
Could we log a warning if we find a fuzzy query with the old format? Maybe the log could be throttled, printed only once per run or only once every x times. This could help in migrating to the new fuzzy format.

I think the PR is missing some tests (or do we already have these tests?)
- The `~` character appears in the query but not as part of a fuzzy match.
- The query contains several fuzzy match expressions. 
- The query contains a mix of old and new style fuzzy matches.

fabriziofortino commented:
@alagodasii CI in this repo fails mostly because it's running with low resources. If you have failures while building the repo locally, please report them.

fabriziofortino committed:
OAK-11603: improve fuzzy conversion

fabriziofortino commented:
@nfsantos 
> Could we log a warning if we find a fuzzy query with the old format? Maybe the log could be throttled, printed only once per run or only once every x times. This could help in migrating to the new fuzzy format.

I added a log.trace call when this conversion happens. I don't think it would be correct to have a warning here because, so far, there is no plan of deprecating/migrating lucene in favour of elasticsearch or a more recent version of lucene. Queries executed against an Elastic index have to be backward compatible with Lucene queries based on version 4.x but not the other way around. A query with `roam~2` would work on Elastic only. We cannot change Lucene to be forward compatible with more recent versions.

> 
> I think the PR is missing some tests (or do we already have these tests?)
> 
> * The `~` character appears in the query but not as part of a fuzzy match.
> * The query contains several fuzzy match expressions.
> * The query contains a mix of old and new style fuzzy matches.

I added all these cases.
